rule-engine/rule-engine-components/src/main/java/org/thingsboard/rule/engine/cache/TbAbstractCacheBasedRuleNode.java

```diff
@@ -0,0 +1,155 @@
+
+import lombok.extern.slf4j.Slf4j;
+import org.thingsboard.rule.engine.api.RuleNodeCacheService;
+import org.thingsboard.rule.engine.api.TbContext;
+import org.thingsboard.rule.engine.api.TbNode;
+import org.thingsboard.rule.engine.api.TbNodeConfiguration;
+import org.thingsboard.rule.engine.api.TbNodeException;
+import org.thingsboard.server.common.data.id.EntityId;
+import org.thingsboard.server.common.data.id.RuleNodeId;
+import org.thingsboard.server.common.data.plugin.ComponentLifecycleEvent;
+import org.thingsboard.server.common.msg.TbMsg;
+import org.thingsboard.server.common.msg.TbMsgMetaData;
+import org.thingsboard.server.common.msg.queue.PartitionChangeMsg;
+import org.thingsboard.server.common.msg.queue.TopicPartitionInfo;
+
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Set;
+import java.util.concurrent.ExecutionException;
+import java.util.function.Consumer;
+import java.util.stream.Collectors;
+
+@Slf4j
+public abstract class TbAbstractCacheBasedRuleNode<C, K> implements TbNode {
+
+    protected static final int DEFAULT_PARTITION = -999999;
+    protected static final TbMsgMetaData EMPTY_META_DATA = new TbMsgMetaData();
+    protected static final String EMPTY_DATA = "";
+
+    protected C config;
+
+    protected final Map<Integer, Set<EntityId>> partitionsEntityIdsMap;
+    protected final Map<EntityId, K> entityIdValuesMap;
+
+    protected TbAbstractCacheBasedRuleNode() {
+        partitionsEntityIdsMap = new HashMap<>();
+        entityIdValuesMap = new HashMap<>();
+    }
+
+    @Override
+    public void init(TbContext ctx, TbNodeConfiguration configuration) throws TbNodeException {
+        this.config = loadRuleNodeConfiguration(configuration);
+        getValuesFromCacheAndSchedule(ctx);
+    }
+
+    @Override
+    public void onMsg(TbContext ctx, TbMsg msg) throws ExecutionException, InterruptedException, TbNodeException {
+        if (getTickMsgType().equals(msg.getType())) {
+            processOnTickMsg(ctx, msg);
+        } else {
+            processOnRegularMsg(ctx, msg);
+        }
+    }
+
+    @Override
+    public void destroy(TbContext ctx, ComponentLifecycleEvent reason) {
+        if (ComponentLifecycleEvent.DELETED.equals(reason)) {
+            if (!partitionsEntityIdsMap.isEmpty()) {
+                partitionsEntityIdsMap.forEach((partition, entityIds) ->
+                        entityIds.forEach(id ->
+                                getCacheIfPresentAndExecute(ctx, cache -> cache.evictTbMsgs(id, partition))));
+            }
+            getCacheIfPresentAndExecute(ctx, cache -> cache.evict(getEntityIdsCacheKey()));
+        }
+        partitionsEntityIdsMap.clear();
+        entityIdValuesMap.clear();
+    }
+
+    @Override
+    public void onPartitionChangeMsg(TbContext ctx, PartitionChangeMsg msg) {
+        Set<Map.Entry<Integer, Set<EntityId>>> currentPartitions = partitionsEntityIdsMap.entrySet();
```

Comment:
Why do we execute all of this if the cache is disabled? It is better to call ```ctx.getRuleNodeCacheService()``` on init and store the cache variable. If the variable is null, we should skip all this logic.

Reply:
I think that we anyway need to execute all of the logic even if the cache is absent in order to do the maps cleanup if the rebalancing happens.